bin/p2-exec/main.go

```diff
@@ -0,0 +1,163 @@
+package main
+
+import (
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"syscall"
+
+	"github.com/square/p2/pkg/cgroups"
+	"github.com/square/p2/pkg/util"
+	"gopkg.in/yaml.v2"
+)
+
+// usage without cgroups:
+//   p2-exec username envdir launchable:cgname myexecutable myargs
+// there must be a PLATFORM_CONFIG_PATH envar (possibly in the envdir) to
+// specify the cgroup parameters for the <launchable>, which is created under
+// <cgname>
+func main() {
+	username := os.Args[1]
+	envDir := os.Args[2]
+	cgParams := os.Args[3]
+
+	newCmd := os.Args[4:]
+	if len(newCmd) == 0 {
+		log.Fatal("No command was passed (args %q)", os.Args)
+	}
+
+	err := loadEnvDir(envDir)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	err = nolimit()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	if platconf := os.Getenv("PLATFORM_CONFIG_PATH"); platconf != "" {
+		err = cgEnter(platconf, cgParams)
+		if err != nil {
+			log.Fatal(err)
+		}
+	}
+
+	err = changeUser(username)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	binPath, err := exec.LookPath(newCmd[0])
+	if err != nil {
+		log.Fatal(err)
+	}
+	err = syscall.Exec(binPath, newCmd, os.Environ())
+	// should never be reached
+	if err != nil {
+		log.Fatalf("Error executing command %q: %s", newCmd, err)
+	}
+}
+
+// loadEnvDir iterates over all files in the given directory and loads them as
+// environment variables, matching the behavior of chpst -e or envdir.
+func loadEnvDir(dir string) error {
+	envFiles, err := ioutil.ReadDir(dir)
+	if err != nil {
+		return err
+	}
+
+	for _, envFile := range envFiles {
+		if envFile.IsDir() {
+			continue
+		}
+		if strings.IndexByte(envFile.Name(), '=') != -1 {
+			return util.Errorf("env file %q cannot contain equals sign in its filename", filepath.Join(dir, envFile.Name()))
```

Comment:
No, just the equals sign because it separates the key from the value in the representation of the environment (which is a list of null-separated strings where each string is a key, an equals sign and a value). Other characters (eg non-printing control chars, unicode) would be strange, but from a technical standpoint, they would work fine.